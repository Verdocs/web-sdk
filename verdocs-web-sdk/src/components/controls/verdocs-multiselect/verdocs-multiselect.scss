@import '../../../theme.scss';

verdocs-multiselect {
  width: 100%;
  display: block;
  margin: 0 0 10px 0;
  box-sizing: border-box;
  font-family: $primary-font;

  .input-wrapper {
    width: 100%;
    position: relative;

    .icon {
      top: 6px;
      right: 6px;
      width: 18px;
      height: 18px;
      z-index: 10;
      position: absolute;
    }
  }

  &.open .input-wrapper .icon {
    transform: rotate(180deg);
  }

  .trigger {
    gap: 3px;
    width: 100%;
    display: flex;
    flex-wrap: wrap;
    max-width: 100%;
    border-radius: 4px;
    flex-direction: row;
    box-sizing: border-box;
    padding: 5px 32px 5px 7px;
    background: $verdocs-grey-4;
    border: 1px solid $border-color;

    &::placeholder {
      opacity: 1;
      color: $light-disabled-color;
    }
  }

  .selected-option {
    color: white;
    font-size: 12px;
    padding: 2px 6px;
    border-radius: 3px;
    display: inline-block;
    background: $verdocs-grey-blue;
  }

  .input-label {
    display: block;
    font-weight: 700;
    font-size: 14px;
    margin: 0 0 4px 0;
    color: $label-color-light;
  }
}

#verdocs-multiselect-menu-items {
  width: 200px;
  z-index: 10000;
  position: absolute;
  border-radius: 5px;
  font-family: $primary-font;
  background: $verdocs-grey-4;
  border: 1px solid $item-border-color;

  label {
    font-size: 13px;
  }
}
