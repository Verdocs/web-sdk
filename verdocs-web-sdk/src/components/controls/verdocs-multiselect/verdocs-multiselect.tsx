import {randomString} from '@verdocs/js-sdk';
import {Component, Prop, Host, h, State, EventEmitter, Event} from '@stencil/core';

export interface IMultiSelectOption {
  label: string;
  value: string;
}

const DownIcon = `<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="size-6"><path stroke-linecap="round" stroke-linejoin="round" d="m19.5 8.25-7.5 7.5-7.5-7.5" /></svg>`;

/**
 * Display a dropdown that allows multiple options to be selected. Note that events "bubble" from the
 * input field to the container, so you can subscribe to the same DOM events (input, blur, etc) that a
 * normal input would emit.
 *
 * ```ts
 * <verdocs-multiselect label="Methods:" value={[]} options={[...options]} onInput={() => {}} />
 * ```
 */
@Component({
  tag: 'verdocs-multiselect',
  styleUrl: 'verdocs-multiselect.scss',
})
export class VerdocsMultiselect {
  /**
   * The label for the field.
   */
  @Prop() label: string = '';

  /**
   * The placeholder for the input element when no options are selected.
   */
  @Prop() placeholder: string = 'Select...';

  /**
   * The options to list.
   */
  @Prop() options: IMultiSelectOption[];

  /**
   * The currently selected options.
   */
  @Prop() selectedOptions: string[] = [];

  @Event({composed: true}) selectionChanged: EventEmitter<{selectedOptions: string[]}>;

  @State() showPicker = false;

  handleToggleOption(e: any, option: IMultiSelectOption) {
    if (e.target.checked) {
      this.selectedOptions = [...this.selectedOptions, option.value];
    } else {
      this.selectedOptions = this.selectedOptions.filter(selected => selected !== option.value);
    }

    this.selectionChanged?.emit({selectedOptions: this.selectedOptions});
  }

  render() {
    const elId = `verdocs-dropdown-${randomString(8)}`;

    return (
      <Host class={this.showPicker ? 'open' : 'closed'}>
        <label>
          {this.label ? <div class="input-label">{this.label + ':'}</div> : <div />}
          <div class="input-wrapper">
            <button id={elId} class="trigger" value="" onClick={() => (this.showPicker = true)}>
              {this.selectedOptions.length === 0 ? (
                <span class="placeholder">{this.placeholder}</span>
              ) : (
                this.selectedOptions.map(option => <span class="selected-option">{this.options.find(opt => opt.value === option)?.label || 'Unknown'}</span>)
              )}
            </button>

            <div class="icon" innerHTML={DownIcon} />

            {this.showPicker && (
              <verdocs-portal anchor={elId} voffset={2} onClickAway={() => (this.showPicker = false)} id="verdocs-multiselect-menu-items">
                {this.options.map((option, i) => (
                  <div class="option" onClick={() => (this.showPicker = false)}>
                    <verdocs-checkbox
                      size="small"
                      id={`verdocs-multi-select-opt-${i}`}
                      value={option.value}
                      checked={this.selectedOptions.includes(option.value)}
                      onInput={(e: any) => this.handleToggleOption(e, option)}
                    />
                    {/*<input*/}
                    {/*  id={`verdocs-multi-select-option-${i}`}*/}
                    {/*  type="checkbox"*/}
                    {/*  value={option.value}*/}
                    {/*  checked={this.selectedOptions.includes(option.value)}*/}
                    {/*  onInput={(e: any) => this.handleToggleOption(e, option)}*/}
                    {/*/>*/}
                    <label htmlFor={`verdocs-multi-select-option-${i}`}>{option.label}</label>
                  </div>
                ))}
              </verdocs-portal>
            )}
          </div>
        </label>
      </Host>
    );
  }
}
