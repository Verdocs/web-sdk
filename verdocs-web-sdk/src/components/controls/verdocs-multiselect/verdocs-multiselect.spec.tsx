import { newSpecPage } from '@stencil/core/testing';
import { VerdocsMultiselect } from './verdocs-multiselect';

describe('verdocs-multiselect', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsMultiselect],
      html: `<verdocs-multiselect options='[]' selectedOptions='[]'></verdocs-multiselect>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with label, placeholder, and options', async () => {
    const options = [
      { label: 'Option 1', value: '1' },
      { label: 'Option 2', value: '2' }
    ];
    const page = await newSpecPage({
      components: [VerdocsMultiselect],
      html: `<verdocs-multiselect label="Test" placeholder="Choose..." options='${JSON.stringify(options)}' selectedOptions='[]'></verdocs-multiselect>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with selected options', async () => {
    const options = [
      { label: 'Option 1', value: '1' },
      { label: 'Option 2', value: '2' }
    ];
    const selectedOptions = ['1'];
    const page = await newSpecPage({
      components: [VerdocsMultiselect],
      html: `<verdocs-multiselect options='${JSON.stringify(options)}' selectedOptions='${JSON.stringify(selectedOptions)}'></verdocs-multiselect>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('emits selectionChanged event on option toggle', async () => {
    const options = [
      { label: 'Option 1', value: '1' },
      { label: 'Option 2', value: '2' }
    ];
    const page = await newSpecPage({
      components: [VerdocsMultiselect],
      html: `<verdocs-multiselect options='${JSON.stringify(options)}' selectedOptions='[]'></verdocs-multiselect>`,
    });

    const spy = jest.fn();
    page.root.addEventListener('selectionChanged', spy);

    // Simulate toggling the first option
    (page.rootInstance as any).handleToggleOption({ target: { checked: true } }, options[0]);
    await page.waitForChanges();

    expect(spy).toHaveBeenCalled();
    expect((page.rootInstance as any).selectedOptions).toContain('1');
  });
});
