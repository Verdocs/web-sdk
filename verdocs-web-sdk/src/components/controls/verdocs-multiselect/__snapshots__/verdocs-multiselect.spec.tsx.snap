// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-multiselect renders with default props 1`] = `
<verdocs-multiselect class="closed">
  <label>
    <div></div>
    <div class="input-wrapper">
      <button class="trigger" id="verdocs-dropdown-4fzzzxjyl" value="">
        <span class="placeholder">
          Select...
        </span>
      </button>
      <div class="icon">
        <svg fill="none" stroke="currentColor" stroke-width="1.5" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
          <path d="m19.5 8.25-7.5 7.5-7.5-7.5" stroke-linecap="round" stroke-linejoin="round"></path>
        </svg>
      </div>
    </div>
  </label>
</verdocs-multiselect>
`;

exports[`verdocs-multiselect renders with label, placeholder, and options 1`] = `
<verdocs-multiselect class="closed">
  <label>
    <div class="input-label">
      Test:
    </div>
    <div class="input-wrapper">
      <button class="trigger" id="verdocs-dropdown-4fzzzxjyl" value="">
        <span class="placeholder">
          Choose...
        </span>
      </button>
      <div class="icon">
        <svg fill="none" stroke="currentColor" stroke-width="1.5" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
          <path d="m19.5 8.25-7.5 7.5-7.5-7.5" stroke-linecap="round" stroke-linejoin="round"></path>
        </svg>
      </div>
    </div>
  </label>
</verdocs-multiselect>
`;

exports[`verdocs-multiselect renders with selected options 1`] = `
<verdocs-multiselect class="closed">
  <label>
    <div class="input-label">
      Test:
    </div>
    <div class="input-wrapper">
      <button class="trigger" id="verdocs-dropdown-4fzzzxjyl" value="">
        <span class="selected-option">
          Option 1
        </span>
      </button>
      <div class="icon">
        <svg fill="none" stroke="currentColor" stroke-width="1.5" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
          <path d="m19.5 8.25-7.5 7.5-7.5-7.5" stroke-linecap="round" stroke-linejoin="round"></path>
        </svg>
      </div>
    </div>
  </label>
</verdocs-multiselect>
`;
